R/boxr_collab.R

```diff
@@ -0,0 +1,259 @@
+    user_id <- NULL
+  }
+  
+  item <-
+    list(
+      type = "folder",
+      id = as.character(dir_id)
+    )
+  
+  accessible_by <-
+    list(
+      type = "user", #  imagine inviting a group
+      id = as.character(user_id),
+      login = login
+    )
+  
+  box_invite(item, accessible_by, role, can_view_path)
+}
+
+#' @rdname box_create_collab_dir
+#' @export
+#' 
+box_create_collab_file <- function(fild_id, user_id, type = "file", login = NULL,
+                                  role = "viewer", can_view_path = FALSE) {
+  # if login is provided, ignore user_id
+  if (!is_void(login)) {
+    user_id <- NULL
+  }
+  
+  item <-
+    list(
+      type = "file",
+      id = as.character(fild_id)
+    )
+  
+  accessible_by <-
+    list(
+      type = "user", #  imagine inviting a group
+      id = as.character(user_id),
+      login = login
+    )
+  
+  box_create_collab(item, accessible_by, role, can_view_path)
+} 
+
+#'
+#'
+box_create_collab <- function(item, accessible_by, role, can_view_path = FALSE) {
+
+  # ref: https://developer.box.com/reference#collaboration-object
+  
+  # validate
+  item_type_legal <- c("file", "folder")
+  
+  acc_type_legal <- c("user", "group")
+  
+  role_legal <- c(
+    "editor", 
+    "viewer", 
+    "previewer", 
+    "uploader", 
+    "previewer uploader", 
+    "viewer uploader",
+    "co-owner", 
+    "owner"
+  )
+  
+  assertthat::assert_that(
+    # item
+    is.list(item),
+    rlang::is_string(item$type),
+    item$type %in% item_type_legal,
+    rlang::is_string(item$id),
+    # accessible_by
+    is.list(accessible_by),
+    rlang::is_string(accessible_by$type),
+    accessible_by$type %in% acc_type_legal,
+    rlang::is_string(accessible_by$id) || rlang::is_string(accessible_by$login),
+    # role
+    rlang::is_string(role),
+    role %in% role_legal,   
+    # can_view_path
+    rlang::is_bool(can_view_path)
+  )
+  
+  # call the API
+  resp <- httr::RETRY(
+    "POST",
+    "https://api.box.com/2.0/collaborations",
+    get_token(),
+    encode = "multipart",
+    body = 
+      jsonlite::toJSON(
+        list(
+          item = item, 
+          accessible_by = accessible_by, 
+          role = role, 
+          can_view_path = can_view_path
+        ),
+        auto_unbox = TRUE
+      ),
+    terminate_on = box_terminal_http_codes()
+  ) 
+  
+  httr::stop_for_status(resp, task = "invite collaborator")
+
+  # TODO: create an S3 class
+  resp <- httr::content(resp)
+  
+  # feedback
+  message(
+    glue::glue(
+      "{resp$created_by$name} ({resp$created_by$login}) has invited",
+      "{resp$accessible_by$name} ({resp$accessible_by$login})",
+      "to collaborate on {resp$item$type} `{resp$item$name}`",
+      "as {resp$role}.",
+      .sep = " "
+    )
+  )
+  
+  invisible(resp)
+}
+
+#' Get the existing collaborations on a file or folder.
+#' 
+#' @param dir_id `integer` ID for Box directory
+#' @param file_id `integer` ID for Box file
+#'
+#' @export
```

Comment:
@return ?

Reply:
Do you think this will be a different return object than `box_create_collab` I assumed it would inherit the  @return value form there...